src/useFetch.ts

```diff
@@ -1,73 +1,130 @@
-import { invariant, isObject } from './utils'
-
-
-export function useFetch<TData = any>(arg1?: useFetchArg1, arg2?: Options | RequestInit): UseFetch<TData> {
+import { HTTPMethod, /* Options, */ UseFetch, FetchCommands, DestructuringCommands, UseFetchResult } from './types'
+import { OptionsNoURLs, OptionsAsFirstParam, OptionsAsFirstParamWithContext, URLOrOptions, UseFetchOptions } from './types'
+import { invariant, isObject, isString, pullOutRequestInit } from './utils'
+
+// type Options = OptionsAsFirstParam | OptionsAsFirstParamWithContext | OptionsNoURLs
+
+// No <Provider url='example.com' />
+function useFetch<TData = any>(url: string, options?: OptionsNoURLs): UseFetch<TData>
+function useFetch<TData = any>(options: OptionsAsFirstParam): UseFetch<TData>
+// With <Provider url='example.com' />
+// options should be extended. In future maybe have options callback to completely overwrite options
+// i.e. useFetch('ex.com', oldOptions => ({ ...newOptions })) to overwrite
+function useFetch<TData = any>(url?: string, options?: OptionsNoURLs): UseFetch<TData>
+function useFetch<TData = any>(options?: OptionsAsFirstParamWithContext): UseFetch<TData>
+
+// TODO: handle context.graphql
+function useFetch<TData = any>(urlOrOptions?: URLOrOptions, optionsNoURLs?: OptionsNoURLs): UseFetch<TData> {
   const context = useContext(FetchContext)
 
-  invariant(!!arg1 && !!context.url, 'The first argument of useFetch is required unless you have a global url setup like: <Provider url="https://example.com"></Provider>')
-
-  let url: string | null = context.url || null
-  let options = {} as { signal?: AbortSignal | null } & RequestInit
-  let onMount = false
-  let baseUrl = ''
-  let method: string = HTTPMethod.GET
-
-  const handleOptions = useCallback((opts: Options & RequestInit) => {
-    if (true) {
-      // take out all the things that are not normal `fetch` options
-      // need to take this out of scope so can set the variables below correctly
-      let { url, onMount, timeout, baseUrl, ...rest } = opts
-      options = { signal: undefined, ...rest }
-    }
-    if (context.url) url = context.url
-    if (opts.url) url = opts.url || context.url || ''
-    if (opts.onMount) onMount = opts.onMount
-    if (opts.method) method = opts.method
-    if (opts.baseUrl) baseUrl = opts.baseUrl
+  // TODO: this needs to be per initial setup below since we need to check urlOrOptions.url OR  urlOrOptions.baseUrl
+  invariant(!!urlOrOptions && !!context.url, 'The first argument of useFetch is required unless you have a global url setup like: <Provider url="https://example.com"></Provider>')
+
+  let url: string = context.url || ''
+  let options: RequestInit = {}
+  let onMount: boolean = false
+  // let timeout: number = 10 // TODO: not implemented
+  // let baseURL: string = ''
+  let method: HTTPMethod = HTTPMethod.GET
+
+  const handleUseFetchOptions = useCallback((useFetchOptions?: UseFetchOptions): void => {
```

Comment:
do we definitely need `useCallback`.  I think you should always be using the [eslint hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks) which checks all this for you and adds the deps.  I find it invaluable.  I could add eslint in a PR if you are open to that but please feel free to say no.  This is your baby

Reply:
@dagda1  Yes add it in please! :) Also, could you help me debug the types portion for the `onMount` useEffect?

Reply:
@dagda1 let's pair on some of this stuff. I think you guys are 8 hours ahead of us. Is it better in the AM for you or PM?

Reply:
@alex-cory this week is an absolute nightmare for me, I'm having to travel from Glasgow to Edinburgh each day and I'm on holiday all next week.  Is there another way I can help in the mean time?

Reply:
of course! We need to get tests out. I got tests for `useFetch` and it using `Provider`. In this PR I will create files for each of the hooks with comments on every test case that I can think of that needs to get tested. I'll try to get that out by today. If you have any time to help implement those, that would be an amazing help. @dagda1 I'll try to get this PR landed today so that you can work on top of master.

Reply:
@alex-cory don't put yourself under pressure.  Do it when you have time.  Open source can become a burden.  You create your own pressure.  Don't feel you have to have anything out by any time.  We're all busy so don't feel you have to do anything by any time.